third_party/move/move-model/src/ty.rs

```diff
@@ -543,6 +545,34 @@ impl Constraint {
                 *a1 = a1.union(*a2);
                 Ok(true)
             },
+            // After the above checks on same type of constraint
+            // Check compatibility between ability and number
+            // This check is needed because sometime the concrete integer type is not available
+            // TODO: check other combination of constraints may be necessary as well.
+            (Constraint::HasAbilities(a1, _), Constraint::SomeNumber(_)) => {
+                let unsupported_abilities = a1.setminus(AbilitySet::PRIMITIVES);
```

Comment:
nit: would be nice if the common duplicate code in these two arms could be refactored out.

Reply:
I was trying to merge these two arms but could not do it because one requires mutable reference and the other one does not. Once we decide we want to generalize this check to different types of constraints, I will refactor them out.